src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1027 @@
+        if let Some(future) = flushed_chunk_future {
+            match future.await {
+                Ok((_, chunk, _)) => {
+                    *flushed_chunk_future = None;
+                    Ok(Some(chunk))
+                }
+                Err(_) => {
+                    // TODO: log + propagate error
+                    *flushed_chunk_future = None;
+                    Ok(None)
+                }
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn try_next_buffer_item(
+        truncation_offset: &mut Option<ReaderTruncationOffsetType>,
+        read_flushed_chunk_future: &mut Option<ReadFlushedChunkFuture>,
+        serde: &LogStoreRowSerde,
+        state_store: &S,
+        buffer: &mut Mutex<SyncedLogStoreBuffer>,
+        table_id: TableId,
+        read_metrics: &KvLogStoreReadMetrics,
+    ) -> StreamExecutorResult<Option<StreamChunk>> {
+        let Some((item_epoch, item)) = buffer.lock().pop_front() else {
+            return Ok(None);
+        };
+        match item {
+            LogStoreBufferItem::StreamChunk {
+                chunk, end_seq_id, ..
+            } => {
+                truncation_offset.replace((item_epoch, Some(end_seq_id)));
+                Ok(Some(chunk))
+            }
+            LogStoreBufferItem::Flushed {
+                vnode_bitmap,
+                start_seq_id,
+                end_seq_id,
+                chunk_id,
+            } => {
+                truncation_offset.replace((item_epoch, Some(end_seq_id)));
+                let serde = serde.clone();
+                let read_metrics = read_metrics.clone();
+                let read_flushed_chunk_fut = read_flushed_chunk(
+                    serde,
+                    state_store.clone(),
+                    vnode_bitmap,
+                    chunk_id,
+                    start_seq_id,
+                    end_seq_id,
+                    item_epoch,
+                    table_id,
+                    read_metrics,
+                )
+                .boxed();
+                *read_flushed_chunk_future = Some(read_flushed_chunk_fut);
+                Self::try_next_flushed_chunk_future(read_flushed_chunk_future).await
+            }
+            LogStoreBufferItem::Barrier { next_epoch, .. } => {
+                // FIXME(kwannoel): Is `next_epoch` correct for truncation??
+                truncation_offset.replace((next_epoch, None));
+                Ok(None)
+            }
+            LogStoreBufferItem::UpdateVnodes(_) => Ok(None),
+        }
+    }
+}
+
+// Write methods
+impl<S: StateStore, LS: LocalStateStore> SyncedKvLogStoreExecutor<S, LS> {
+    async fn write_barrier(
+        state_store: &mut LS,
+        serde: &LogStoreRowSerde,
+        barrier: Barrier,
+        metrics: &mut KvLogStoreMetrics,
+        truncation_offset: Option<ReaderTruncationOffsetType>,
+        seq_id: &mut SeqIdType,
+        buffer: &mut Mutex<SyncedLogStoreBuffer>,
```

Comment:
The current code, the `Mutex<SyncedLogStoreBuffer>` always has the mutable reference. Therefore, it's likely that we don't have to wrap the buffer with the `Mutex`.